node/pkg/guardiansigner/generatedsigner.go

```diff
@@ -0,0 +1,63 @@
+package guardiansigner
+
+import (
+	"crypto/ecdsa"
+	"crypto/rand"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	ethcrypto "github.com/ethereum/go-ethereum/crypto"
+)
+
+// The GeneratedSigner is a signer that is intended for use in tests. It uses the private
+// key supplied to GenerateSignerWithPrivatekeyUnsafe, or defaults to generating a random
+// private key if no private key is supplied.
+type GeneratedSigner struct {
+	privateKey *ecdsa.PrivateKey
+}
+
+func NewGeneratedSigner(key *ecdsa.PrivateKey) (*GeneratedSigner, error) {
+	if key == nil {
+		privateKey, err := ecdsa.GenerateKey(crypto.S256(), rand.Reader)
+		return &GeneratedSigner{privateKey: privateKey}, err
+	} else {
+		return &GeneratedSigner{privateKey: key}, nil
+	}
+
+}
+
+func (gs *GeneratedSigner) Sign(hash []byte) (sig []byte, err error) {
+	// Sign the hash
+	sig, err = crypto.Sign(hash, gs.privateKey)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign wormchain address: %w", err)
+	}
+
+	return sig, nil
+}
+
+func (gs *GeneratedSigner) PublicKey() (pubKey ecdsa.PublicKey) {
+	return gs.privateKey.PublicKey
+}
+
+func (gs *GeneratedSigner) Verify(sig []byte, hash []byte) (valid bool, err error) {
+	recoveredPubKey, err := ethcrypto.SigToPub(hash, sig)
+
+	if err != nil {
+		return false, err
+	}
+
+	// Need to use gs.privateKey.Public() instead of PublicKey to ensure
+	// the returned public key has the right interface for Equal() to work.
+	fsPubkey := gs.privateKey.Public()
```

Comment:
I'm going to leave this as is for now. I remember fooling around with this when I was writing it. Casting the way you suggested fails, because the compiler complains that `gs.PublicKey() is not an interface`.

What I don't understand is why passing in `gs.PublicKey()` (type `ecdsa.PublicKey`) doesn't work, but `gs.privateKey.Public()` does (type `crypto.PublicKey`), despite the `crypto.PublicKey` being casted to `ecdsa.PublicKey` in the `Equal()` function [here](https://cs.opensource.google/go/go/+/refs/tags/go1.23.2:src/crypto/ecdsa/ecdsa.go;l=77). I might be missing something obvious.